File: AoC2022/src/day02.jl
module day02

using InlineTest
using Match
const TEST_STRING = """A Y
B X
C Z"""


@testset "day02" begin
  @test solve(IOBuffer(TEST_STRING)) == (
    15,
    12
  )
end

@enum Moves Rock=1 Paper=2 Sissors=3

function win_score(x, y)
  val = 0
  if x == y
    val = 1
  elseif (Int(y) - Int(x) == 1) || (Int(x) - Int(y) == 2)
    val = 2
  end
  val * 3
end

function parse_move(move)
  @match move begin
    "A" => Rock
    "B" => Paper
    "C" => Sissors
    "X" => Rock
    "Y" => Paper
    "Z" => Sissors
  end
end

function score(moves)
  win_score(moves...) + Int(moves[2])
end


function load(file)
  parse(x) = parse_move.(split(x))
  parse.(readlines(file))
end


function problem1(A)
  sum(score.(A))
end

const beats = Dict(Rock=>Paper, Paper=>Sissors, Sissors=>Rock)
const ties = Dict(Rock=>Rock, Paper=>Paper, Sissors=>Sissors)
const loses = Dict(Rock=>Sissors, Paper=>Rock, Sissors=>Paper)
const predict_moves = Dict(Rock=>loses, Paper=>ties, Sissors=>beats)

function predict_move(move)
  predict_moves[move[2]][move[1]]
end

function problem2(A)
  B = deepcopy(A)
  for i=1:length(B)
    B[i][2] = predict_move(B[i])
  end
  sum(score.(B))
end

function solve(io::IO)
  A = load(io)
  (
    problem1(A), 
    problem2(A)
  )
end

end

File: 2022/AoC2022_cpp/src/day02/day02.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>

#include "utils.h"

enum RPS { Rock = 1, Paper = 2, Sissors = 3 };

struct MovePair {
  RPS elf;
  RPS me;
};

RPS get_move_that_beats(RPS rps) {
  switch (rps) {
  case Rock:
    return Paper;
  case Paper:
    return Sissors;
  case Sissors:
    return Rock;
  }
  throw;
  return Rock;
}

int win_score(const MovePair &move) {
  int score = 0;
  if (move.elf == move.me) {
    score = 1;
  } else if (get_move_that_beats(move.elf) == move.me) {
    score = 2;
  }
  return score * 3;
}

int score(const MovePair &move) { return win_score(move) + move.me; }

int problem1(std::vector<MovePair> moves) {
  std::vector<int> scores;
  for (auto move : moves) {
    scores.push_back(score(move));
  }
  return std::reduce(scores.begin(), scores.end());
}

RPS get_move_that_loses(RPS rps) {
  switch (rps) {
  case Rock:
    return Sissors;
  case Paper:
    return Rock;
  case Sissors:
    return Paper;
  }
  throw;
  return Rock;
}

MovePair get_new_move(const MovePair &move) {
  switch (move.me) {
  case Rock:
    return MovePair{move.elf, get_move_that_loses(move.elf)};
  case Paper:
    return MovePair{move.elf, move.elf};
  case Sissors:
    return MovePair{move.elf, get_move_that_beats(move.elf)};
  }
  throw;
  return MovePair{Rock, Rock};
}

int problem2(std::vector<MovePair> moves) {
  std::vector<int> scores;
  for (auto move : moves) {
    scores.push_back(score(get_new_move(move)));
  }
  return std::reduce(scores.begin(), scores.end());
}

RPS stoRPS(const std::string &in) {
  if (in == "X" || in == "A") {
    return Rock;
  } else if (in == "Y" || in == "B") {
    return Paper;
  } else if (in == "Z" || in == "C") {
    return Sissors;
  }
  throw;
}

std::vector<MovePair> parse(const std::string &filename) {
  std::vector<MovePair> out;
  load(filename, [&out](const std::string &line) {
    auto move = split(line, ' ');
    out.push_back(MovePair{stoRPS(move[0]), stoRPS(move[1])});
  });
  return out;
}

int main(int argc, char *argv[]) {
  std::string filename(argv[1]);
  auto moves = parse(filename);
  std::cout << problem1(moves) << std::endl;
  std::cout << problem2(moves) << std::endl;
  return 0;
}